src/pstore/pstore.c

```diff
@@ -0,0 +1,582 @@
+            r = errno;
+        }
+    }
+    else
+    if (arg_storage == PSTORE_STORAGE_JOURNAL)
+    {
+        //sd_journal_send("MESSAGE=%s", dc->de.d_name, NULL);
+        if (dc->is_binary)
+        {
+            //WITH_BINARY= but how to know end of binary blob if contained in 0 terminated string?
+        }
+        else
+        {
+            r = sd_journal_send("MESSAGE=File %s:\n%s", dc->de.d_name, dc->contents, NULL);
+        }
+        /* NOTE: If journald not running, r always 0 */
+        remove_file = (r == 0);
+    }
+
+    /* If file copied properly, remove it from pstore */
+    if ((r == 0) && remove_file && arg_unlink)
+    {
+        unlink(ifdpath);
+    }
+
+    return r;
+}
+
+static void write_dmesg (const char *dmesg, ssize_t size, const char *id) {
+    if ((dmesg != NULL) && (size > 0))
+    {
+        _cleanup_free_ char *ofdpath;
+        int ofd;
+
+        log_info("Record ID %s\n", id);
+
+        ofdpath = path_join(arg_archivedir, id, "dmesg.txt");
+
+        ofd = open(ofdpath, O_CREAT|O_TRUNC|O_WRONLY, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH/*0644*/);
+        if (ofd < 0)
+        {
+            log_error_errno(errno, "open(%s): %m\n", ofdpath);
+        }
+        else
+        {
+            ssize_t wr;
+            wr = write(ofd, dmesg, size);
+            if (wr != size)
+            {
+                log_error_errno(errno, "write(%s): %m\n", ofdpath);
+            }
+        }
+        close(ofd);
+    }
+}
+
+/*
+ * PStore content rule handling
+ */
+typedef struct rules_s
+{
+    void (*handler)(struct rules_s *rule, dirent_ct *dc);
+    const char *handlerName;
+    const char *pattern;
+} rules_t;
+static void dmesg_handler (struct rules_s *rule, dirent_ct *dc);
+static void move_handler (struct rules_s *rule, dirent_ct *dc);
+rules_t rules[] =
+{
+    /* from Linux/fs/pstore/inode.c
+     * the %s is record->psi->name (eg efi, erst)
+     * the %llu is record->id
+     */
+    { dmesg_handler, "dmesg", "dmesg-*", },     /* PSTORE_TYPE_DMESG: "dmesg-%s-%llu%s"     */
+    { move_handler, "console", "console-*", },  /* PSTORE_TYPE_CONSOLE: "console-%s-%llu"   */
+    { move_handler, "pmsg", "pmsg-*", },        /* PSTORE_TYPE_PMSG: "pmsg-%s-%llu"         */
+    { move_handler, "ftrace", "ftrace-*", },    /* PSTORE_TYPE_FTRACE: "ftrace-%s-%llu"     */
+    { move_handler, "mce", "mce-*", },          /* PSTORE_TYPE_MCE: "mce-%s-%llu"           */
+    { move_handler, "ppcrtas", "rtas-*", },     /* PSTORE_TYPE_PPC_RTAS: "rtas-%s-%llu"     */
+    { move_handler, "ppcof", "powerpc-ofw-*", },/* PSTORE_TYPE_PPC_OF: "powerpc-ofw-%s-%llu"*/
+    { move_handler, "ppccommon", "powerpc-common-*", }, /* PSTORE_TYPE_PPC_COMMON: "powerpc-common-%s-%llu" */
+    { move_handler, "ppcopal", "powerpc-opal-*", },     /* PSTORE_TYPE_PPC_OPAL: "powerpc-opal-%s-%llu"     */
+    { move_handler, "unknown", "unknown-*", },      /* PSTORE_TYPE_UNKNOWN: "unknown-%s-%llu"   */
+};
```

Comment:
why do we need this table btw? why not just copy over all files in "blanket" style? i.e. why only some here?

Reply:
In general, the move_handler is for file types I don't understand. Today I currently only have experience with the DMESG flavor. So for the types that are at least recognized by Linux, files are copied over, in the spirit of your question.
In the case of DMESG, I do extra processing on the files in order to make those individual files into something more user friendly. By offering this "handler" table, it provides a framework for others to do something more interesting (beyond copying) for the other types, if warranted.

Reply:
hmm, it appears to me it would be better to just blanket copy everything bit-by-bit, and process it on display, in this case.

Reply:
This ties in nicely with @yuwata question about the need to sort. In the case of processing DMESG file types in the pstore, I do need to sort the files (in the case of DMESG, sort in reverse lexigraphical) in order to reconstruct the dmesg.
I likely don't understand exactly what you mean by "blanket copy everything and process it on display", but if that is to mean just copy the files and reconstruct dmesg on the fly, then I can not reconstruct dmesg. The reasons being: file order as returned by readdir() is not sorted, and I need files sorted in order to: find boundaries of unrelated dmesgs, and to reconstruct a particular dmesg. For example, on one of our servers, the readdir of /sys/fs/pstore results in:
dmesg-erst-6706482473678667778
dmesg-erst-6706482473678667779
dmesg-erst-6706404902550781953
dmesg-erst-6706404902550781954
dmesg-erst-6706404902550781955
dmesg-erst-6706482473678667777
(which is an artifact of the way the non-volatile storage recycled). Point being I can not rely on any kind of ordering of the pstore entries.
which when sorted looks like:
# ls -al /sys/fs/pstore
total 0
drwxr-x--- 2 root root     0 Jun 25 12:08 .
drwxr-xr-x 2 root root     0 Jun 25 12:08 ..
-r--r--r-- 1 root root 17714 Jun 24 23:28 dmesg-erst-6706404902550781953
-r--r--r-- 1 root root 17757 Jun 24 23:28 dmesg-erst-6706404902550781954
-r--r--r-- 1 root root 17716 Jun 24 23:28 dmesg-erst-6706404902550781955
-r--r--r-- 1 root root 17738 Jun 25 10:57 dmesg-erst-6706482473678667777
-r--r--r-- 1 root root 17751 Jun 25 10:57 dmesg-erst-6706482473678667778
-r--r--r-- 1 root root 17705 Jun 25 10:57 dmesg-erst-6706482473678667779
The above shows two distinct dmesgs, each requiring 3 pstore entries. When I reconstruct the dmesg, after sorting the files, for filenames that are "similar" (by dropping the last 7 chars), then those files belong to the same dmesg. While I could use timestamps (and perhaps somebody will deem that an appropriate enhancement in the future), you still have the same "similar" problem as each pstore file can have a timestamp that is not identical.
Anyway, all this just to reconstruct DMESG; I've no idea what the other types might want. So that is why I put in the handler infrastructure, and that is also why I need sorting.

Reply:
hmm, so, i still wonder: why do we need the table here? wouldn't a simpler logic be: "process dmesg specially, and everything else the same"?